python/ql/lib/semmle/python/frameworks/SqlAlchemy.qll

```diff
@@ -113,15 +113,25 @@ module SqlAlchemy {
      */
     abstract class InstanceSource extends DataFlow::LocalSourceNode { }
 
+    /**
+     * join-ordering helper for ConnectionConstruction char-pred -- without this would
+     * start with _all_ `CallCfgNode` and join those with `MethodCallNode` .. which is
+     * silly
+     */
+    pragma[noinline]
+    private DataFlow::MethodCallNode connectionConstruction_helper() {
+      result.calls(Engine::instance(), ["begin", "connect"])
+      or
+      result.calls(instance(), ["connect", "execution_options"])
+    }
+
     private class ConnectionConstruction extends InstanceSource, DataFlow::CallCfgNode {
       ConnectionConstruction() {
-        this = classRef().getACall()
-        or
-        this.(DataFlow::MethodCallNode).calls(Engine::instance(), ["begin", "connect"])
+        // without the `pragma[only_bind_out]` we would start with joining
+        // `API::Node.getACall` with `CallCfgNode` which is not optimal
+        this = pragma[only_bind_out](classRef()).getACall()
         or
-        this.(DataFlow::MethodCallNode).calls(instance(), "connect")
-        or
-        this.(DataFlow::MethodCallNode).calls(instance(), "execution_options")
+        this = connectionConstruction_helper()
       }
     }
```

Comment:
Could we, instead of the helper function, do `pragma[only_bind_out]` on `Engine::instance()` and `instance()`?
(And would that read nicer?)

Reply:
If you want the join-order pipeline to start with those, then that's certainly an option.

Reply:
I can't remember if I tried that exact combo, but I think I had a hard time making it work without the helper predicate.

> would that read nicer?

In this specific case, I'm leaning towards no (my personal taste is that code with `pragma[only_bind_out/into]` generally reads poorly)

Reply:
> If you want the join-order pipeline to always start with those, then that's certainly an option.

I think that is what we what want here; identified instances are rather specific, whereas being a `MethodCallNode` or a `CallCfgNode` is rather general.

Reply:
> > would that read nicer?
> 
> In this specific case, I'm leaning towards no (my personal taste is that code with `pragma[only_bind_out/into]` generally reads poorly)

But is that just because we are unfamiliar with the pragmas? I generally think that helper predicates read like "optimisation clutter" unless they have clear names _and_ semantics. I think, that in this case, the intent of "find the instances first" is clearly communicated by the pragmas and no so clearly by `connectionConstruction_helper`.

Reply:
```codeql
        this = classRef().getACall()
        or
        this.(DataFlow::MethodCallNode).calls(Engine::instance(), ["begin", "connect"])
        or
        this.(DataFlow::MethodCallNode).calls(instance(), "connect")
        or
        this.(DataFlow::MethodCallNode).calls(instance(), "execution_options")
```

No what I mean is that my eyes can easily read the code above, and understand what is going on. When trying to understand what code is doing, I translate the pragmas into "performance stuff", so it's not a matter of trying to understand what they do, but the fact that they're all over the code which hinders my eyes to easily read it. As a slightly more extreme example try reading the code below :joy:

```codeql
        pragma[perf_stuff](this) = pragma[perf_stuff](pragma[perf_stuff](classRef()).getACall())
        or
        pragma[perf_stuff](this).(DataFlow::MethodCallNode).calls(pragma[perf_stuff](Engine::instance()), pragma[perf_stuff](["begin", "connect"]))
        or
        pragma[perf_stuff](this).(DataFlow::MethodCallNode).calls(pragma[perf_stuff](instance()), pragma[perf_stuff]("connect"))
        or
        pragma[perf_stuff](this).(DataFlow::MethodCallNode).calls(pragma[perf_stuff](instance()), pragma[perf_stuff]("execution_options"))
```

Reply:
Ah, if we need that many pragmas, then I agree 🙂  But in such a case, you probably want a `inline_late` `no_magic` combo or something. Or there will be a natural helper predicate catching the underlying reasoning.
My hope is that we will generally just need a few targeted pragmas. If that turns out to not be true, we either need very short pragma names (say `[->]` and `[<-]`), editor support, or a different mechanism (like falling back to helper predicates)...

Reply:
I was just trying to prove a point about readability :wink: Seeing as the intention behind this discussion is aligning for future work, I think the takeaway is: you need to make judgements on a case-by-case basis :blush: